sql/src/main/java/io/crate/execution/ddl/tables/AlterTableOperation.java

```diff
@@ -236,6 +284,211 @@ private Session session() {
+            () -> revertAllocation(sourceIndexName),
+            () -> CompletableFuture.completedFuture(-1L)
+        ));
+        return ChainableActions.run(actions);
+    }
+
+    private void validateForResize(String indexName, int targetNumberOfShards) {
+        final int currentNumberOfShards = clusterService.state().metaData().index(indexName).getNumberOfShards();
+        if (currentNumberOfShards == targetNumberOfShards) {
+            throw new IllegalArgumentException(
+                String.format("Table/partition is already allocated <%d> shards", currentNumberOfShards));
+        } else if (targetNumberOfShards > currentNumberOfShards) {
+            throw new IllegalArgumentException("Increasing the number of shards is not supported");
+        }
+
+        final Boolean blocks_write = clusterService.state().metaData().index(indexName)
+            .getSettings().getAsBoolean(INDEX_BLOCKS_WRITE_SETTING.getKey(), false);
+        if (!blocks_write) {
+            throw new IllegalStateException("Table/Partition needs to be at a read-only state." +
+                                               " Use 'ALTER table ... set (\"blocks.write\"=true)' and retry");
+        }
+    }
+
+    private DiscoveryNode getNodeForResize() {
+        assert !clusterService.state().nodes().getDataNodes().isEmpty() : "No Data nodes available " +
+                                                                          "for index resize operation";
+        return clusterService.state().nodes().getDataNodes().valuesIt().next();
+    }
+
+    private CompletableFuture<Long> restrictAllocationToNode(String nodeName, String... indices) {
+        Settings settings = Settings.builder()
+            .put(INDEX_ROUTING_REQUIRE_GROUP_PREFIX + "._name", nodeName)
+            .build();
+
+        UpdateSettingsRequest request = new UpdateSettingsRequest(settings, indices);
+        request.indicesOptions(IndicesOptions.lenientExpandOpen());
+
+        FutureActionListener<UpdateSettingsResponse, Long> listener
+            = new FutureActionListener<>(r -> (r.isAcknowledged()) ? 0L : -1L);
+
+        transportUpdateSettingsAction.execute(request, ActionListener.wrap(response -> {
+            if (response.isAcknowledged()) {
+                activeShardsObserver.waitForActiveShards(request.indices(),
+                    ActiveShardCount.DEFAULT,
+                    request.ackTimeout(), shardsAcked -> {
+                        if (!shardsAcked && logger.isInfoEnabled()) {
+                            logger.info("[{}] indexes reallocation is in progress, " +
+                                        "but the operation timed out while waiting for " +
+                                        "enough shards to be back on green state. Timeout={}. " +
+                                        "Consider increasing the timeout or adding nodes to the cluster.",
+                                request.indices(), request.timeout());
```

Comment:
I think we should either rollback the operation and log a warning or do nothing by best-effort (thus the shrink operation may fail).

Reply:
decided to bail out